### PR TITLE
Add IOUSimilarity layer to handle masking.

### PR DESCRIPTION
**Pull request by tanzhenyu**
# What does this PR do?
IOU with masking

### CODE REVIEW

File: keras_cv/layers/object_detection/iou_similarity.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_cv import bounding_box
+
+
+class IouSimilarity(tf.keras.layers.Layer):
+    """Computes a lookup table matrix containing the ious for a given pair of boxes.
```
Comment (ianstenbit) [inline keras_cv/layers/object_detection/iou_similarity.py:21]: I think "pair of boxes" should be "pair of lists of boxes"

Comment (ianstenbit) [inline keras_cv/layers/object_detection/iou_similarity.py:21]: Maybe /s/lookup table matrix/similarity matrix

I think it would be good if the docstring refers to what's returned with the same phrase throughout (currently it calls it both a lookup table matrix and a lookup vector and an iou_lookup_table

File: keras_cv/layers/object_detection/iou_similarity.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    out of the boundary, i.e., all their coodinates are less than 0. Users can pad ground truth
+    boxes with -1. in order for it to be masked here.
+
+    Args:
+      boxes1: a list of bounding boxes in 'corners' format. Can be batched or unbatched.
+      boxes2: a list of bounding boxes in 'corners' format. This should match the rank and
+        shape of boxes1.
+      bounding_box_format: a case-insensitive string.
+        For detailed information on the supported format, see the
+        [KerasCV bounding box documentation](https://keras.io/api/keras_cv/bounding_box/formats/).
```
Comment (ianstenbit) [inline keras_cv/layers/object_detection/iou_similarity.py:38]: Should we add arg descriptors for box1_mask and box2_mask?

File: keras_cv/layers/object_detection/iou_similarity.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    Returns:
+      iou_lookup_table: a vector containing the pairwise ious of boxes1 and
+        boxes2.
+    """
+
+    def __init__(
+        self,
+        box1_format,
```
Comment (ianstenbit) [inline keras_cv/layers/object_detection/iou_similarity.py:47]: (Just curious) -- in what sort of use cases would we expect the two input boxes to be in different formats?

File: keras_cv/layers/object_detection/iou_similarity.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      boxes2: a list of bounding boxes in 'corners' format. This should match the rank and
+        shape of boxes1.
+      bounding_box_format: a case-insensitive string.
+        For detailed information on the supported format, see the
+        [KerasCV bounding box documentation](https://keras.io/api/keras_cv/bounding_box/formats/).
+
+    Returns:
+      iou_lookup_table: a vector containing the pairwise ious of boxes1 and
+        boxes2.
+    """
```
Comment (ianstenbit) [inline keras_cv/layers/object_detection/iou_similarity.py:43]: Should we have an example usage for this layer?

I think that might be useful in explaining the meaning of the masking behavior, but maybe that's overkill

File: keras_cv/layers/object_detection/iou_similarity.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_cv import bounding_box
+
+
+class IouSimilarity(tf.keras.layers.Layer):
```
Comment (LukeWood) [inline keras_cv/layers/object_detection/iou_similarity.py:20]: Lets remove references to "Similarity" here as thats a somewhat overloaded term.  IoU is more precise from a naming perspective.

File: keras_cv/layers/object_detection/iou_similarity.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    ):
+        super().__init__(**kwargs)
+        self.box1_format = box1_format
+        self.box2_format = box2_format
+        self.mask_val = mask_val
+        self.box1_mask = box1_mask
+        self.box2_mask = box2_mask
+        self.built = True
+
+    def call(self, boxes1, boxes2, boxes1_mask=None, boxes2_mask=None):
```
Comment (LukeWood) [inline keras_cv/layers/object_detection/iou_similarity.py:62]: Whats the specific use case of accepting explicit masks into call() instead of masking with -1s?

File: keras_cv/layers/object_detection/iou_similarity.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        elif self.box1_mask:
+            boxes1_mask = tf.less(tf.reduce_max(boxes1, axis=-1, keepdims=True), 0.0)
+            background_mask = boxes1_mask
+        else:
+            boxes2_mask = tf.less(tf.reduce_max(boxes2, axis=-1, keepdims=True), 0.0)
+            background_mask = tf.logical_or(
+                tf.zeros(tf.shape(boxes2)[:-1], dtype=tf.bool),
+                tf.transpose(boxes2_mask, perm),
+            )
+        iou_lookup_table = tf.where(background_mask, mask_val_t, result)
```
Comment (LukeWood) [inline keras_cv/layers/object_detection/iou_similarity.py:91]: Any reason for this to be a layer instead of a utility?  

I wouldn't expect this to be plugged anywhere else into the API in a configurable way (i.e. as a constructor parameter for another component), so I think it makes a lot of sense to keep it as a function.

**Review by LukeWood — CHANGES_REQUESTED** [review]
Personally I think `compute_iou()` fits the construct of a function better than a class.  Here is my reasoning:

- users will never configure this themselves; there is one way and one way only to compute IoU
- any parameters we DO need to configure are on a component-level


As such, I think we should upstream returning -1 for background class boxes to:
https://github.com/keras-team/keras-cv/blob/master/keras_cv/bounding_box/iou.py#L20

instead of creating a new component.

**Comment by LukeWood** [comment]
Thanks for the PR!  Returning -1 IoU for background classes is a much better pattern!

I think IoU is a better fit for a function rather than a component

**Comment by tanzhenyu** [comment]
> Thanks for the PR! Returning -1 IoU for background classes is a much better pattern!
> 
> I think IoU is a better fit for a function rather than a component

I originally wrote it as function, but then calling this in several places carrying information such as "mask_val", "box_mask", "box_format" seems pretty heavy to user, so I switched to layer, users can create it once, use it everywhere.
I'm fine with a function as well. WDYT?

**Comment by LukeWood** [comment]
> > Thanks for the PR! Returning -1 IoU for background classes is a much better pattern!
> > I think IoU is a better fit for a function rather than a component
> 
> I originally wrote it as function, but then calling this in several places carrying information such as "mask_val", "box_mask", "box_format" seems pretty heavy to user, so I switched to layer, users can create it once, use it everywhere. I'm fine with a function as well. WDYT?

are all of the arguments present required?  I think this influences the answer to the questions.

I don't see a ton of use cases for dis-joint bounding box formats, and do we need to provide masks manually?  Or can we craft them based on class_id=background_class?

**Comment by tanzhenyu** [comment]
> > > Thanks for the PR! Returning -1 IoU for background classes is a much better pattern!
> > > I think IoU is a better fit for a function rather than a component
> > 
> > 
> > I originally wrote it as function, but then calling this in several places carrying information such as "mask_val", "box_mask", "box_format" seems pretty heavy to user, so I switched to layer, users can create it once, use it everywhere. I'm fine with a function as well. WDYT?
> 
> are all of the arguments present required? I think this influences the answer to the questions.
> 
> I don't see a ton of use cases for dis-joint bounding box formats, and do we need to provide masks manually? Or can we craft them based on class_id=background_class?

So the behavior is turn on/off the masking. The layer will automatically compute mask for users, with the assumption that padded boxes are with default values less than 0 -- that seems to be a good assumption so far.
Otherwise, we'd have to ask users to compute mask everytime, which is more flexible, but bloats the model code

**Comment by LukeWood** [comment]
> So the behavior is turn on/off the masking. The layer will automatically compute mask for users, with the assumption that padded boxes are with default values less than 0 -- that seems to be a good assumption so far.


If its just one parameter then we can support:

```
compute_iou(boxes1, boxes2, bounding_box_format, use_masking=True):
```

any make masking the default.  If this is the case I think its a better fit as a utility than a layer.  Let me know if there are cases missing in this API.

**Comment by tanzhenyu** [comment]
> > So the behavior is turn on/off the masking. The layer will automatically compute mask for users, with the assumption that padded boxes are with default values less than 0 -- that seems to be a good assumption so far.
> 
> If its just one parameter then we can support:
> 
> ```
> compute_iou(boxes1, boxes2, bounding_box_format, use_masking=True):
> ```
> 
> any make masking the default. If this is the case I think its a better fit as a utility than a layer. Let me know if there are cases missing in this API.

That's fine with me. (Though we might need to support masking for boxes1 only, or for boxes2 only, or both. I'm not seeing that pattern yet)

**Comment by LukeWood** [comment]
> > > So the behavior is turn on/off the masking. The layer will automatically compute mask for users, with the assumption that padded boxes are with default values less than 0 -- that seems to be a good assumption so far.
> > 
> > 
> > If its just one parameter then we can support:
> > ```
> > compute_iou(boxes1, boxes2, bounding_box_format, use_masking=True):
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > any make masking the default. If this is the case I think its a better fit as a utility than a layer. Let me know if there are cases missing in this API.
> 
> That's fine with me. (Though we might need to support masking for boxes1 only, or for boxes2 only, or both. I'm not seeing that pattern yet)

I'm fairly sure we'll always mask out the -1s, so lets try it with a utility and we can always add a parameter like:

```
`masking='boxes1'` or `masking='boxes2'` down the line if a need arises
```

**Comment by tanzhenyu** [comment]
#870

**Comment by tanzhenyu** [comment]
Closing this PR as the FasterRCNN PR already deals with all the comments above.

**Comment by LukeWood** [comment]
Greatl thanks!
